### PR TITLE
Make min-blob-size online configurable

### PR DESCRIPTION
**Pull request by v01dstar**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -95,9 +95,13 @@ if (WITH_TITAN_TESTS OR WITH_TITAN_TOOLS)
     set(CMAKE_REQUIRED_FLAGS "--std=c++11")
     CHECK_CXX_SOURCE_COMPILES("
 #include <atomic>
-std::atomic<uint64_t> x(0);
+struct A {
```
Comment (Connor1996) [inline CMakeLists.txt:98]: why change

File: src/blob_storage.h [hunk 2]
```diff
@@ -53,7 +54,7 @@ class BlobStorage {
 
   TitanCFOptions cf_options() {
     auto _cf_options = cf_options_;
-    _cf_options.blob_run_mode = blob_run_mode_.load();
+    cf_options_.UpdateMutableOptions(mutable_cf_options_);
```
Comment (Connor1996) [inline src/blob_storage.h:57]: `_cf_options` is not changed

File: src/db_impl.cc [hunk 3]
```diff
@@ -1099,49 +1100,66 @@ Options TitanDBImpl::GetOptions(ColumnFamilyHandle* column_family) const {
   return options;
 }
 
+Status TitanDBImpl::ExtractTitanCfOptions(
```
Comment (Connor1996) [inline src/db_impl.cc:1103]: As it never returns non-okay status, why not use `changed` as return value

File: CMakeLists.txt (hunk 1 above)
Comment (v01dstar) [reply CMakeLists.txt:98]: This was for conditionally linking `libatomic`,  atomic ops over structs larger than 16 bytes require the bin link with `libatomic`. Turns out its not a easy thing to do in `rust-rocksdb`. I now use `mutex` over lock free, changing config is not a frequent op and in the code it already acquire the mutex anyway.

File: src/db_impl.cc [hunk 4]
```diff
@@ -1099,49 +1100,66 @@ Options TitanDBImpl::GetOptions(ColumnFamilyHandle* column_family) const {
...
+  // Make sure base db's SetOptions success before setting Titan's.
+  if (titan_options_changed) {
     uint32_t cf_id = column_family->GetID();
     {
       MutexLock l(&mutex_);
       assert(cf_info_.count(cf_id) > 0);
       TitanColumnFamilyInfo& cf_info = cf_info_[cf_id];
-      cf_info.titan_table_factory->SetBlobRunMode(blob_run_mode);
-      cf_info.mutable_cf_options.blob_run_mode = blob_run_mode;
+      cf_info.titan_table_factory->SetMutableCFOptions(mutable_cf_options);
```
Comment (Connor1996) [inline src/db_impl.cc:1158]: Is the mutex same as the mutex held by `SetMutableCFOptions`

Comment (v01dstar) [reply src/db_impl.cc:1158]: Yes, it is the db wide mutex.

Comment (v01dstar) [reply src/db_impl.cc:1158]: https://github.com/v01dstar/titan/blob/7085173919bd6ce906f68435c94ff8dd8ac75d91/src/db_impl.cc#L456

File: CMakeLists.txt (hunk 1 above)
Comment (v01dstar) [reply CMakeLists.txt]: https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html More readings about why we need libatomic.

File: src/db_impl.cc (hunk 4 above)
Comment (Connor1996) [reply src/db_impl.cc:1158]: Then `SetMutableCFOptions` would lock an already locked mutex?

Comment (v01dstar) [reply src/db_impl.cc:1158]: I use `mutext_->AssertHeld()`, right?

Comment (v01dstar) [reply src/db_impl.cc:1158]: TableFactory uses the same mutex, so it is `mutex->AssertHeld()`, BlobStorage uses its own mutex, so it is `MutexLock l(&(blob_storage_)mutex_)`

Comment (Connor1996) [reply src/db_impl.cc:1158]: Oh, I misread

**Review by Connor1996 — APPROVED** [review]
LGTM

File: src/db_impl.cc (hunk 3 above)
Comment (tonyxuqqi) [reply src/db_impl.cc:1103]: The caller below seems not checking the status but just the `changed` parameter

File: src/db_impl.cc [hunk 5]
```diff
@@ -1099,49 +1100,66 @@ Options TitanDBImpl::GetOptions(ColumnFamilyHandle* column_family) const {
...
-                       column_family->GetName().c_str(),
-                       blob_run_mode_string.c_str());
-      }
-      opts.erase(p);
-    }
-  }
+  MutableTitanCFOptions mutable_cf_options =
+      cf_info_.at(column_family->GetID()).mutable_cf_options;
+  bool titan_options_changed = false;
+  ExtractTitanCfOptions(column_family, opts, mutable_cf_options,
```
Comment (tonyxuqqi) [inline src/db_impl.cc:1143]: Here the status is not checked.

**Review by tonyxuqqi — APPROVED** [review]
One minor issue. Others LGTM.

File: src/db_impl.cc (hunk 5 above)
Comment (v01dstar) [reply src/db_impl.cc]: Done.

**Pull request merged**